OpenSim/Actuators/VandenBogert2011Muscle.cpp

```diff
@@ -0,0 +1,1234 @@
+                                                          fiberLength,
+                                                          projFiberLength,
+                                                          normalizedValues);
+    return projFiberVelocity;
+};
+
+
+//------------------------------------------------------------------------------
+double VandenBogert2011Muscle::projFibVelToFiberVelocity(double projFibVel,
+                                                         double fiberLength,
+                                                         double projFiberLength,
+                                                         bool normalizedValues=0)
+                                                         const
+{
+// Note that this normalized by fiber optimal length (not by max velocity)
+    double optimalFiberLength = getOptimalFiberLength();
+
+    if (normalizedValues)
+    {
+        fiberLength = fiberLength * getOptimalFiberLength();
+        projFibVel = projFibVel * optimalFiberLength;
+        projFiberLength = projFiberLength * optimalFiberLength;
+    }
+
+    double fiberVelocity = 0;
+
+    if (projFibVel != 0)
+    {
+        fiberVelocity = projFibVel * cos(projFiberLength / fiberLength);
+    }
+    else { fiberVelocity = 0; }
+
+    if (normalizedValues)
+    {
+        fiberVelocity = fiberVelocity / optimalFiberLength;
+    };
+
+    return fiberVelocity;}
+
+//------------------------------------------------------------------------------
+double VandenBogert2011Muscle::projFibVelToFiberVelocity
+        (double projFibVel, double projFibLength, bool normalizedValues=0)
+                        const {
+
+    //overload method when fiberLength is not known/provided
+
+    double fiberLength=projFibLenToFiberLength(fiberLength ,normalizedValues);
+    double fiberVelocity = projFibVelToFiberVelocity(projFibVel, fiberLength,
+                                                     projFibLength,
+                                                     normalizedValues);
+
+    return fiberVelocity;
+};
+
+
+
+//------------------------------------------------------------------------------
+VandenBogert2011Muscle::ImplicitResidual
+VandenBogert2011Muscle::calcImplicitResidual(const SimTK::State& s,
+                                             double projFibVelNorm_guess,
+                                             double activdot_guess,
+                                             double u, int returnJacobians)
+
+
+{
+    // Overload method for state as parameters
+
+    double muscleLength = getLength(s);
+
+    //double activ =getActivations(s); //Does not work because info not implemented
+    //double activ = getStateVariableValues(s,"activation");  // does not work
+    //double fiberLength = getStateVariableValues(s,"fiber_length");
+
+    SimTK::Vector stateVec =getStateVariableValues(s);
+    double fiberLength =stateVec[0];
+    double activ =stateVec[1];
```

Comment:
You would ideally use `getActivation(s)`, etc. but you can also use `getStateVariableValue(s, "activation")` (no `s`).